internal/pkg/service/appsproxy/http/router.go

```diff
@@ -143,8 +143,10 @@ func (r *Router) CreateHandler() http.Handler {
 		httpHandler := handler.getHTTPHandler()
 
 		if modified || httpHandler == nil {
-			httpHandler = r.createDataAppHandler(req.Context(), config)
-			handler.setHTTPHandler(httpHandler)
+			handler.setHTTPHandler(func() http.Handler {
+				return r.createDataAppHandler(req.Context(), config)
+			})
+			httpHandler = handler.getHTTPHandler()
```

Comment:
Thanks, it is better.

Some more improvements could be made here so that it is all under one lock, but let's leave that for another iteration.